src/main/kotlin/com/mdrobnak/lalrpop/psi/ext/LpTypeRef.kt

```diff
@@ -0,0 +1,71 @@
+package com.mdrobnak.lalrpop.psi.ext
+
+import com.intellij.extapi.psi.ASTWrapperPsiElement
+import com.intellij.lang.ASTNode
+import com.mdrobnak.lalrpop.psi.*
+import com.mdrobnak.lalrpop.psi.util.isRefMut
+import com.mdrobnak.lalrpop.psi.util.lifetimeOrInfer
+import org.rust.lang.core.psi.ext.childrenWithLeaves
+import org.rust.lang.core.psi.ext.elementType
+
+abstract class LpTypeRefMixin(node: ASTNode) : ASTWrapperPsiElement(node), LpTypeRef {
+    override fun resolveType(arguments: List<NonterminalGenericArgument>): String {
+        return when (childrenWithLeaves.first().elementType) {
+            // tuple
+            LpElementTypes.LPAREN -> this.typeRefList.joinToString(prefix = "(", postfix = ")") {
+                it.resolveType(arguments)
+            }
+            // array
+            LpElementTypes.LBRACKET -> "[" + this.typeRefList.first().resolveType(arguments) + "]"
+            // FIXME: not sure if this is supposed to be the type of another symbol,
+            //  but would assume so given the context requires a type while a symbol is passed.
+            LpElementTypes.POUND -> this.symbol?.resolveType(arguments) ?: "()"
+
+            // reference
+            LpElementTypes.AND -> "&${this.lifetimeOrInfer} ${if (this.isRefMut) "mut" else ""} ${
+                this.typeRefList[0].resolveType(arguments)
```

Comment:
I think indexing into the list may fail if the user has only partially typed the type.